core/lib/spree/event.rb

```diff
@@ -52,51 +110,78 @@ def fire(event_name, opts = {})
+    # Still, you can add new subscriptions to the same event and they'll be
+    # called if the event is fired:
+    #
+    # @example
+    #   Spree::Event.subscribe('foo') { do_something }
+    #   Spree::Event.unsubscribe 'foo'
+    #   Spree::Event.subscribe('foo') { do_something_else }
+    #   Spree::Event.fire 'foo' # `do_something_else` will be called, but
+    #   # `do_something` won't
+    #
+    # @param [String, Symbol, Spree::Event::Listener] name_or_subscriber the
+    # event name as a string or the subscription object. Take into account that
+    # if the deprecated {Spree::Event::Adapters::ActiveSupportNotifications}
+    # adapter is used, the subscription object will be a
+    # {ActiveSupport::Notifications::Fanout::Subscribers::Timed} object.
     #
     # @example Unsubscribe a single subscription
-    #   subscription = Spree::Event.fire 'order_finalized'
+    #   subscription = Spree::Event.subscribe('order_finalized') { do_something
+    #   }
     #   Spree::Event.unsubscribe(subscription)
     # @example Unsubscribe all `order_finalized` event subscriptions
     #   Spree::Event.unsubscribe('order_finalized')
-    # @example Unsubscribe an event by name with explicit prefix
-    #   Spree::Event.unsubscribe('order_finalized.spree')
-    def unsubscribe(subscriber)
-      name_or_subscriber = subscriber.is_a?(String) ? normalize_name(subscriber) : subscriber
+    def unsubscribe(name_or_subscriber, adapter: default_adapter)
+      name_or_subscriber = normalize_name(name_or_subscriber)
       adapter.unsubscribe(name_or_subscriber)
     end
 
-    # Lists all subscriptions currently registered under the ".spree"
-    # namespace. Actual implementation is delegated to the adapter
+    # Lists all subscriptions.
     #
-    # @return [Hash] an hash with event names as keys and arrays of subscriptions
-    #  as values
+    # Actual implementation is delegated to the adapter.
+    #
+    # @return [Hash<<String,Regexp>,Spree::Event::Listener>] an hash with
+    # patterns as keys and arrays of subscriptions as values. Take into account
+    # that the deprecated {Spree::Event::Adapters::ActiveSupportNotifications}
+    # adapter will map to
+    # {ActiveSupport::Notifications::Fanout::Subscribers::Timed} instances
+    # instead.
     #
     # @example Current subscriptions
     #  Spree::Event.listeners
-    #    # => {"order_finalized.spree"=> [#<ActiveSupport...>],
-    #      "reimbursement_reimbursed.spree"=> [#<ActiveSupport...>]}
-    def listeners
-      adapter.listeners_for(listener_names)
+    #    # => {"order_finalized"=> [#<Spree::Event::Listener...>],
+    #         "reimbursement_reimbursed"=> [#<Spree::Event::Listener...>]}
+    def listeners(adapter: default_adapter)
+      if legacy_adapter(adapter)
+        adapter.listeners_for(listener_names)
+      else
+        init = Hash.new { |h, k| h[k] = [] }
+        adapter.listeners.each_with_object(init) do |listener, map|
```

Comment:
I have the feeling Ruby is providing some handy shortcuts for this. For example, I think we can avoid creating the init variable and just use `adapter.listeners.each_with_object(Hash.new([]))`, isn't it?

Reply:
Using `Hash.new` with a default value binds the same instance to every key. So if you use a mutable object like an array, you can get into serious troubles :sweat:. E.g.:

```ruby
hash = Hash.new([])
hash[:a] << 1
hash[:b] # => [1]
```

:boom:

Reply:
Ops, you are right. I didn't think about that.